packages/next/src/client/components/use-action-queue.ts

```diff
@@ -50,5 +50,27 @@ export function useActionQueue(
       actionQueue.dispatch(action, setState)
   }
 
-  return isThenable(state) ? use(state) : state
+  // When navigating to a non-prefetched route, then App Router state will be
+  // blocked until the server responds. We need to transfer the `_debugInfo`
+  // from the underlying Flight response onto the top-level promise that is
+  // passed to React (via `use`) so that the latency is accurately represented
+  // in the React DevTools.
+  const stateWithDebugInfo = useMemo(() => {
+    if (isThenable(state)) {
+      const debugInfo: Array<unknown> = []
+      const promiseWithDebugInfo = Promise.resolve(state).then((asyncState) => {
+        if (asyncState.debugInfo !== null) {
+          debugInfo.push(...asyncState.debugInfo)
+        }
+        return asyncState
+      }) as Promise<AppRouterState> & { _debugInfo?: Array<unknown> }
+      promiseWithDebugInfo._debugInfo = debugInfo
+      return promiseWithDebugInfo
+    }
+    return state
+  }, [state])
+
+  return isThenable(stateWithDebugInfo)
+    ? use(stateWithDebugInfo)
+    : stateWithDebugInfo
```

Comment:
Under what circumstances is this not a thenable? It would end up dropping the information in those cases.

Reply:
When the route tree is prefetched, in which case it's able to construct a tree from the cache